tutorials/e2e-genai-langchain-app/README.md

```diff
@@ -0,0 +1,83 @@
+
+
+
+1. If needed, git clone https://github.com/GoogleCloudPlatform/ai-on-gke
+1. Install cluster and kuberay operator
+    1. cd `gke-platform`
+    1. Edit `variables.tf` with your GCP settings. Make sure you change `project_id` and `cluster_name`.
+    1. Run `terraform init`
+    1. Run `terraform apply`
+1. Configure credentials to point to the cluster: `gcloud container clusters get-credentials <cluster-name> --location=<region>`
+1. Install Ray on GKE
+    1. cd `ray-on-gke/user`
+    1. Edit variables.tf with your GCP settings. Make sure you set: `project_id`, `namespace`, `service_account`.
+    1. Note the namespace setting. All microservices in this sample will be deployed to this same namespace for simplicity.
+    1. Run `terraform init`
+    1. Run `terraform apply`
+1. Install Jupyter on GKE
+    1. cd `jupyter-on-gke`
+    1. Edit `variables.tf` with your GCP settings. Make sure that you set `project_id`, `project_number`, `namespace`. Use the same namespace as above.
+    1. Configure higher resource limits and guarantees. In `jupyter_config/config.yaml` change the following:
+        ```yaml
+        singleuser:
+        cpu:
+            limit: 1
+            guarantee: .5
+        memory:
+            limit: 4G
+            guarantee: 1G
+        ```
+    1. Run terraform init
+    1. Run terraform apply
+1. Experiment with the model in Jupyter Notebook
+    1. Get the address of your Jupyter hub from the Services.page:
+![Jupyter Hub IP](./open_jupyter.png). 
+    1. Use default credentials: `user1` and `password` to open notebook. You can configure IAP by following the steps in [here](../../jupyter-on-gke/README.md).
+    1. From JupyterHub open this notebook: `https://raw.githubusercontent.com/GoogleCloudPlatform/ai-on-gke/main/tutorials/langchain/nb1.ipynb` and run it step by step
+    1. The first section shows how to run the model directly
+    1. The second section shows how to do the same using `Ray Serve`.
+1. We used Jupyter Notebook to experiment, but now let's build the Flask backend that calls into `Ray Serve`.
+    1. Observe `model.py`: it loads the model and creates `Ray.Serve` function that uses `Langchain` library to run two nested prompts.
+    1. Observe `main.py`: it uses Flask framework to create API route that calls into `Ray.Serve` endpoint
+    1. Containerize and deploy the backend image to the registry. Do these steps from `backend` directory:
+        ```bash
+        PROJECT_ID=<YOUR_PROJECT_ID>
+        # configure GCR
+        gcloud auth configure-docker
+        # build the image
+        docker build -t hf-lc-ray:latest .
+        # tag the image for GCR
+        docker tag hf-lc-ray:latest gcr.io/${PROJECT_ID}/hf-lc-ray:latest
+        # push the image to GCR
+        docker push gcr.io/${PROJECT_ID}/hf-lc-ray:latest
+        ```
+    1. Deploy backend to the cluster. Open `src/backend/deploy.yaml` and change `PROJECT_ID` to your project. Then run
+        ```bash
```

Comment:
Optional: just do this from the command line
```suggestion
    1. Deploy backend to the cluster. Then run
        ```bash
        sed -i "s/YOUR_PROJECT/${PROJECT_ID}/" src/backend/deploy.yaml 
```

Reply:
Added as suggestion here and for frontend.